File: cpp/01.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <numeric>
#include <vector>

#include "util.h"

int main(int argc, char* argv[]) {
  std::vector<int> left, right;
  
  ASSERT_EQ(argc, 2) << "Usage: " << argv[0] << " <data file>";

  {
    std::ifstream data{argv[1]};
    while (data) {
      data >> left;
      data >> right;
    }
  }

  ASSERT_EQ(left.size(), right.size());

  std::sort(left.begin(), left.end());
  std::sort(right.begin(), right.end());

  size_t dist = std::inner_product(
      left.begin(), 
      left.end(), 
      right.begin(), 
      0, 
      std::plus<int>{}, 
      [](int a, int b) {
        return std::abs(a - b);
      });
  std::cout << dist << std::endl;

  return 0;
}

File: rust/01.rs
use std::env;
use std::process;
use std::fs::File;
use std::io::BufReader;
use std::io::BufRead;

fn main() {
    if env::args().count() < 2 {
        eprintln!("Usage: {} <data file>", env::args().nth(0).unwrap());
        process::exit(1);
    }

    let (mut left, mut right): (Vec<i32>, Vec<i32>) = BufReader::new(File::open(env::args().nth(1).unwrap()).unwrap()).lines()
        .map(|e| e.unwrap().split_whitespace()
            .map(|e| e.parse::<i32>().unwrap())
            .collect::<Vec<i32>>())
        .map(|e| (e[0], e[1]))
        .unzip();

    if left.len() != right.len() {
        eprintln!("left size: {}", left.len());
        eprintln!("right size: {}", right.len());
        process::exit(2);
    }

    left.sort();
    right.sort();
    
    let dist = left.into_iter().zip(right.into_iter())
        .map(|(l, r)| (l - r).abs())
        .reduce(|acc, e| acc + e).unwrap();

    println!("{dist}");
}
